src/sql/workbench/contrib/notebook/browser/models/notebookModel.ts

```diff
@@ -1028,4 +1062,158 @@ export class NotebookModel extends Disposable implements INotebookModel {
 		this._contentChangedEmitter.fire(changeInfo);
 	}
 
+	findNext(): Thenable<NotebookRange> {
+		if (this._findArray && this._findArray.length !== 0) {
+			if (this._findIndex === this._findArray.length - 1) {
+				this._findIndex = 0;
+			} else {
+				++this._findIndex;
+			}
+			return Promise.resolve(this._findArray[this._findIndex]);
+		} else {
+			return Promise.reject(new Error('no search running'));
+		}
+	}
+
+	findPrevious(): Thenable<NotebookRange> {
+		if (this._findArray && this._findArray.length !== 0) {
+			if (this._findIndex === 0) {
+				this._findIndex = this._findArray.length - 1;
+			} else {
+				--this._findIndex;
+			}
+			return Promise.resolve(this._findArray[this._findIndex]);
+		} else {
+			return Promise.reject(new Error('no search running'));
+		}
+	}
+
+	find(exp: string, maxMatches?: number): Promise<NotebookRange> {
+		this._findArray = new Array<NotebookRange>();
+		this._findIndex = 0;
+		this._onFindCountChange.fire(this._findArray.length);
+		if (exp) {
+			return new Promise<NotebookRange>((resolve) => {
+				const disp = this.onFindCountChange(e => {
+					resolve(this._findArray[0]);
+					disp.dispose();
+				});
+				this._startSearch(exp, maxMatches);
+			});
+		} else {
+			return Promise.reject(new Error('no expression'));
+		}
+	}
+
+	public get findMatches(): NotebookFindMatch[] {
+		let findMatches: NotebookFindMatch[] = [];
+		this._findArray.forEach(element => {
+			findMatches = findMatches.concat(new NotebookFindMatch(element, null));
+		});
+		return findMatches;
+	}
+
+	public get findArray(): NotebookRange[] {
+		return this.findArray;
+	}
+
+	private _startSearch(exp: string, maxMatches: number = 0): void {
+		let searchFn = (cell: ICellModel, exp: string): NotebookRange[] => {
+			let findResults: NotebookRange[] = [];
+			let cellVal = cell.source;
+			let index: number;
+			let start: number;
+			let end: number;
+			if (cellVal) {
+				if (typeof cellVal === 'string') {
```

Comment:
This could really use some comments explaining what each step is doing. Tests would be great too - testing the logic here shouldn't be too bad.

Reply:
I'll work on adding tests.